gpcontrib/arenadata_toolkit/src/arenadata_toolkit_worker.c

```diff
@@ -0,0 +1,361 @@
+#include "arenadata_toolkit_guc.h"
+#include "bloom_set.h"
+#include "tf_shmem.h"
+
+#define TOOLKIT_BINARY_NAME "arenadata_toolkit"
+
+typedef struct
+{
+	Oid			dbid;
+	Name		dbname;
+	bool		get_full_snapshot_on_recovery;
+}	tracked_db_t;
+
+/* flags set by signal handlers */
+static volatile sig_atomic_t got_sighup = false;
+static volatile sig_atomic_t got_sigterm = false;
+
+void arenadata_toolkit_main(Datum);
+
+/* parse array of GUCs, find desired and analyze it */
+static bool
+is_db_tracked(ArrayType *array)
+{
+	bool		is_tracked = false;
+	Datum	   *elems;
+	bool	   *nulls;
+	int			nelems;
+
+	deconstruct_array(array, TEXTOID, -1, false, 'i',
+					  &elems, &nulls, &nelems);
+	for (int i = 0; i < nelems; i++)
+	{
+		char	   *s;
+		char	   *name;
+		char	   *value;
+
+		if (nulls[i])
+			continue;
+
+		s = TextDatumGetCString(elems[i]);
+		ParseLongOption(s, &name, &value);
+
+		if (!value)
+		{
+			free(name);
+			continue;
+		}
+
+		if (strcmp(name, "arenadata_toolkit.tracking_is_db_tracked") == 0 &&
+			strcmp(value, "t") == 0)
+		{
+			is_tracked = true;
+			break;
+		}
+
+		free(name);
+		if (value)
+			free(value);
+		pfree(s);
+	}
+
+	return is_tracked;
+}
+
+static bool
+full_snapshot_on_recovery(ArrayType *array)
+{
+	bool		take_snapshot = false;
+	bool		found = false;
+	Datum	   *elems;
+	bool	   *nulls;
+	int			nelems;
+
+	deconstruct_array(array, TEXTOID, -1, false, 'i',
+					  &elems, &nulls, &nelems);
+
+	for (int i = 0; i < nelems; i++)
+	{
+		char	   *s;
+		char	   *name;
+		char	   *value;
+
+		if (nulls[i])
+			continue;
+
+		s = TextDatumGetCString(elems[i]);
+		ParseLongOption(s, &name, &value);
+
+		if (!value)
+		{
+			free(name);
+			continue;
+		}
+
+		if (strcmp(name, "arenadata_toolkit.tracking_snapshot_on_recovery") == 0)
+		{
+			found = true;
+			if (strcmp(value, "t") == 0)
+				take_snapshot = true;
+			break;
+		}
+
+		free(name);
+		if (value)
+			free(value);
+		pfree(s);
+	}
+
+	if (!found)
+		take_snapshot = get_full_snapshot_on_recovery;
+
+	return take_snapshot;
+}
+
+/*
+ * Signal handler for SIGTERM
+ * Set a flag to let the main loop to terminate, and set our latch to wake
+ * it up.
+ */
+static void
+tracking_sigterm(SIGNAL_ARGS)
+{
+	int			save_errno = errno;
+
+	got_sigterm = true;
+	if (MyProc)
+		SetLatch(&MyProc->procLatch);
+
+	errno = save_errno;
+}
+
+/*
+ * Signal handler for SIGHUP
+ * Set a flag to tell the main loop to reread the config file, and set
+ * our latch to wake it up.
+ */
+static void
+tracking_sighup(SIGNAL_ARGS)
+{
+	int			save_errno = errno;
+
+	got_sighup = true;
+	if (MyProc)
+		SetLatch(&MyProc->procLatch);
+
+	errno = save_errno;
+}
+
+static List*
+get_tracked_dbs()
+{
+	Relation	rel;
+	SysScanDesc scan;
+	HeapTuple	tup;
+	List		*tracked_dbs = NIL;
+	tracked_db_t *trackedDb;
+
+	rel = heap_open(DbRoleSettingRelationId, AccessShareLock);
+	scan = systable_beginscan(rel, InvalidOid, false, NULL, 0, NULL);
+	while (HeapTupleIsValid(tup = systable_getnext(scan)))
```

Comment:
Why do it manually so difficult? It can be done much easier through SPI, for example, like [this](https://github.com/RekGRpth/pg_quota/blob/2c52103e07a8f4765492281422f05d3028975cd4/pg_quota.c#L423-L458).